Source/WebCore/Modules/webaudio/MediaStreamAudioSourceGStreamer.cpp

```diff
@@ -65,14 +46,25 @@ void MediaStreamAudioSource::consumeAudio(AudioBus& bus, size_t numberOfFrames)
         m_caps = adoptGRef(gst_audio_info_to_caps(&m_info));
     }
 
-    size_t size = GST_AUDIO_INFO_BPS(&m_info) * bus.numberOfChannels() * numberOfFrames;
-    auto buffer = adoptGRef(gst_buffer_new_allocate(nullptr, size, nullptr));
-
+    auto buffer = adoptGRef(gst_buffer_new());
     GST_BUFFER_PTS(buffer.get()) = toGstClockTime(mediaTime);
     GST_BUFFER_FLAG_SET(buffer.get(), GST_BUFFER_FLAG_LIVE);
 
-    copyBusData(bus, buffer.get(), muted());
+    for (size_t channelIndex = 0; channelIndex < bus.numberOfChannels(); ++channelIndex) {
+        auto& channel = *bus.channel(channelIndex);
+        auto dataSize = sizeof(float) * channel.length();
+
+        bus.ref();
+        gst_buffer_append_memory(buffer.get(), gst_memory_new_wrapped(GST_MEMORY_FLAG_READONLY, channel.mutableData(), dataSize, 0, dataSize, &bus, reinterpret_cast<GDestroyNotify>(+[](gpointer data) {
+            auto bus = reinterpret_cast<AudioBus*>(data);
+            bus->deref();
```

Comment:
I don't like reffing and unreffing. I think maybe a GRefPtr would be better but you have to create it in the heap and delete it in the GDestroyNotify, so it would be the same thing in the end. You can keep it, just a thought.